### PR TITLE
fix!: Retry config

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

- https://github.com/awslabs/aws-sdk-swift/issues/1735

## Description of changes

- Adds back the previously lost config `maxAttempts` that allows SDK users to configure number of retries the SDK should make without having to directly manipulate the `retryStategyOptions` config option.
- Also: Fixes the config defaults provider and codegen for it, such that the `maxAttempts` and `awsRetryMode` config values set by the user actually get used when we resolve the default `retryStrategyOptions` config.

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/AWSClientConfigDefaultsProvider.swift [hunk 1]
```diff
@@ -99,17 +99,30 @@ public class AWSClientConfigDefaultsProvider {
         return resolvedRetryMode!
```
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/AWSClientConfigDefaultsProvider.swift:99]: GH diff is being a bit wonky here. I just added `private static func maxAttempts`, and changed the function signature of `func retryStrategyOptions` to take in arguments (the top level config `awsRetryMode` and `maxAttempts`) at call site.

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Config/AWSDefaultClientConfiguration.swift [hunk 2]
```diff
@@ -31,9 +31,19 @@ public protocol AWSDefaultClientConfiguration {
```
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Config/AWSDefaultClientConfiguration.swift:31]: Adds back the lost `maxAttempts` option back to config. Welcome back to our poor option that finally found its way back home.

File: Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Plugins/DefaultAWSClientPlugin.swift [hunk 3]
```diff
@@ -15,10 +15,13 @@ public class DefaultAWSClientPlugin: Plugin {
     }
```
Comment (sichanyoo) [inline Sources/Core/AWSClientRuntime/Sources/AWSClientRuntime/Plugins/DefaultAWSClientPlugin.swift:15]: Now that defaults provider for retry strategy options actually takes in user-exposed config now, let's use em.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSHttpProtocolServiceClient.kt [hunk 4]
```diff
@@ -67,7 +67,7 @@ class AWSHttpProtocolServiceClient(
                     ConfigProperty(
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/AWSHttpProtocolServiceClient.kt:67]: In initializers that take in arguments for `awsRetryMode` and `maxAttempts`, default value for retry strategy options should be defaults provider function call using those arguments. 

In initializer that doesn't take in arguments for `awsRetryMode` and `maxAttempts` but needs to set retry strategy options value to something that isn't nil (e.g., the region initializer), we can't pass in any values to defaults provider method call.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/config/AWSDefaultClientConfiguration.kt [hunk 5]
```diff
@@ -39,5 +39,6 @@ class AWSDefaultClientConfiguration : ClientConfiguration {
             { it.format("\$N.retryMode()", AWSClientRuntimeTypes.Core.AWSClientConfigDefaultsProvider) },
```
Comment (sichanyoo) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/config/AWSDefaultClientConfiguration.kt:39]: Adds the new config value in codegen side.

**Review by sichanyoo — COMMENTED** [review]
Comments for reviewers

**Review by jbelkins — APPROVED** [review]

**Pull request merged**